mpi4jax/_src/jax_compat.py

```diff
@@ -36,7 +25,42 @@ def check_jax_version():
     from jax.interpreters.xla import Token
 
 
+if jax_version >= versiontuple("0.3.15"):
+    # abstract eval needs to return effects
+    # see https://github.com/google/jax/issues/11620
+    from functools import wraps
+    from jax.interpreters import mlir
+    from jax._src.lax import control_flow as lcf
```

Comment:
So what do you think? We should probably pin `jax` to a maximum version and let that be bumped by dependabot?

Reply:
Yeah I would be absolutely in favour of this.
It's going to be annoying though.

We could eventually build a Github action that bumps the version on master and if it works it also bumps it on the latest tagged release, and it it passes, it tags a new .post release that also supports the new jax...

Reply:
Alternative: no upper constraint, but if we error on load of mpi4jax we print an informative message saying that the most recent version that is know to work is XXX, and that either mpi4jax is outdated or it is not yet compatible with the current version.

Reply:
I don't mind triggering a manual release every couple of weeks. If we find it's annoying after a while we can still automate it like you suggested.

Reply:
Actually there's a problem. Assume people have jax and jaxlib 0.3.15 installed and run `pip install mpi4jax`. This will then pull `jax<=0.3.14`, but since we don't depend on jaxlib it will result in a broken state.

We could either start depending on jaxlib with the same version requirements as jax, or go for your other suggestion (emit a warning on import). Perhaps the warning is cleaner / less invasive?

Reply:
but in that case they'll get a quite nice jax error.

```python
In [1]: import jax

File ~/Documents/pythonenvs/netket_env/lib/python3.9/site-packages/jax/_src/lib/__init__.py:89, in check_jaxlib_version(jax_version, jaxlib_version, minimum_jaxlib_version)
     85 if _jaxlib_version > _jax_version:
     86   msg = (f'jaxlib version {jaxlib_version} is newer than and '
     87          f'incompatible with jax version {jax_version}. Please '
     88          'update your jax and/or jaxlib packages.')
---> 89   raise RuntimeError(msg)
     91 return _jaxlib_version

RuntimeError: jaxlib version 0.3.15 is newer than and incompatible with jax version 0.3.14. Please update your jax and/or jaxlib packages.
```

Reply:
well, unfortunately this suggests updating jax which would end up in a broken mpi4jax state anyway...

Reply:
Maybe version bounds + warning?

Reply:
Sure but if people then follow the suggestion from the error and do `pip install -U jax` we didn't gain anything :smile:

Reply:
do you want to hop on a zoom call to discuss it quickly?

Reply:
meet.google.com/amu-pdcm-wyf

Reply:
Yea guess we missed each other. But thinking more about it I think there are only 2 clean solutions:

1. Depend on jax and jaxlib and accept that some people will have their (incompatible) GPU jaxlib downgraded to a CPU one.
2. Do not depend on either and emit a warning during import if the jax version is higher than the tested one. In most cases we will be upwards compatible so it should be fine, otherwise we give the user the tools to deal with it if they encounter errors.

Now the question is: Do we care more about being able to run out of the box, or do we care more about not messing with people's environments?

Reply:
Option 1 is risky. I'd prefer users to handle their jaxlib on their own, IMO. At least until jax people do the long announced jaxlib splitting where the GPU becomes an extra package/extension that is installed independently.

I prefer option 2.

The question is: do we print a warning always, or only if something crashes/fails to load?

Reply:
I would always print it, maybe with an environment variable to silence it. Too hard to anticipate all possible failure modes.

Reply:
Fine by me